src/storage/v2/replication/replication_client.cpp

```diff
@@ -38,10 +38,16 @@ ReplicationStorageClient::ReplicationStorageClient(::memgraph::replication::Repl
 void ReplicationStorageClient::UpdateReplicaState(Storage *storage, DatabaseAccessProtector db_acc) {
   auto &replStorageState = storage->repl_storage_state_;
 
-  auto hb_stream{client_.rpc_client_.Stream<replication::HeartbeatRpc>(main_uuid_, storage->uuid(),
-                                                                       replStorageState.last_durable_timestamp_,
-                                                                       std::string{replStorageState.epoch_.id()})};
-  const auto replica = hb_stream.AwaitResponse();
+  // default-constructed HeartbeatRpc
+  replication::HeartbeatRes replica;
+
+  // stream should be destroyed so that RPC lock is released before taking engine lock
+  {
+    auto hb_stream{client_.rpc_client_.Stream<replication::HeartbeatRpc>(main_uuid_, storage->uuid(),
+                                                                         replStorageState.last_durable_timestamp_,
+                                                                         std::string{replStorageState.epoch_.id()})};
+    replica = hb_stream.AwaitResponse();
+  }
```

Comment:
```suggestion
  replication::HeartbeatRes replica = std::invoke([&]{
    // stream should be destroyed so that RPC lock is released 
    // before taking engine lock
    auto hb_stream = client_.rpc_client_.Stream<replication::HeartbeatRpc>(
      main_uuid_, 
      storage->uuid(),
      replStorageState.last_durable_timestamp_,
      std::string{replStorageState.epoch_.id()}
    );
    return hb_stream.AwaitResponse();
  });
```